perception/autoware_tensorrt_yolox/src/utils.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "autoware/tensorrt_yolox/utils.hpp"
+
+namespace autoware::tensorrt_yolox
+{
+
+std::vector<std::pair<uint8_t, int>> runLengthEncoder(const cv::Mat & image)
+{
+  std::vector<std::pair<uint8_t, int>> compressed_data;
+  const int rows = image.rows;
+  const int cols = image.cols;
+  compressed_data.emplace_back(image.at<uint8_t>(0, 0), 0);
+  for (int i = 0; i < rows; ++i) {
+    for (int j = 0; j < cols; ++j) {
+      uint8_t current_value = image.at<uint8_t>(i, j);
+      if (compressed_data.back().first == current_value) {
+        ++compressed_data.back().second;
+      } else {
+        compressed_data.emplace_back(current_value, 1);
+      }
+    }
+  }
+  return compressed_data;
+}
+
+cv::Mat runLengthDecoder(const std::vector<uint8_t> & rle_data, const int rows, const int cols)
+{
+  cv::Mat mask(rows, cols, CV_8UC1, cv::Scalar(0));
+  int idx = 0;
+  int step = sizeof(uint8_t) + sizeof(int);
+  int nb_pixels = 0;
+  for (size_t i = 0; i < rle_data.size(); i += step) {
+    uint8_t value;
+    int length;
+    std::memcpy(&value, &rle_data[i], sizeof(uint8_t));
+    std::memcpy(&length, &rle_data[i + sizeof(uint8_t)], sizeof(int));
```

Comment:
I believe this code works with no problem, but let me discuss a semantic (intension of the writer) topic.
In my understanding, the value used in the `[]` operator is indices, which include the size of each element implicitly (i.e., if a case of `std::vector<int> int_vec`, the distance of `&int_vec[i]`  and `&int_vec[i+1]` would be 4), while the `sizeof` operator returns the actual byte size of the specified object. 
To distinguish "index" and "address" (written in bytes), how about the following modification?:
```suggestion
    std::memcpy(&length, &rle_data[i + 1], sizeof(int));    // under the condition that we know rle_data[i] only consume 1 element of the vector
```
```suggestion
    std::memcpy(&length, reinterpret_cast<uint8_t*>(reinterpret_cas<void*>(&rle_data[i]) + sizeof(uint8_t)), sizeof(int));     // straightforward address calculation
```

This point is kind of my preference, so if you feel something is wrong, please let me know!

Reply:
@manato Thank you for your suggestion.

> the value used in the [] operator is indices, 

Here the value in [] is memory index of `int length` in `data`, which is started after `uint8_t value (the semantic class)`, so that [] has 1byte (sizeof(uint8_t)) padding from beggining of each encoded pattern. The result should be the same as your suggestion.

Reply:
@badai-nguyen 
Thank you for your consideration. 
Yes, I believe the result should be the same regardless of my suggestion being introduced. My concern was the mixture of "index" and "byte-size-oriented operation". Since your latest commit uses `+1` manipulate index, my concern has been gone, anyway.